File: Day45/Part7/tower_defense.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

struct Point {
    int x, y;
};

struct Enemy {
    int path_idx;
    int hp;
    int speed;
};

double dist(Point p1, Point p2) {
    return sqrt(pow(p1.x - p2.x, 2) + pow(p1.y - p2.y, 2));
}

bool check(const vector<Point>& towers, const vector<vector<Point>>& paths, const vector<Enemy>& enemies, int radius, int damage, int interval) {
    for(const auto& e : enemies) {
        const auto& path = paths[e.path_idx];
        int hp = e.hp;
        int speed = e.speed;
        
        int max_time = path.size() / speed; // Approx
        
        for(int t=0; t < max_time; t += interval) {
            int idx = t * speed;
            if(idx >= path.size()) break;
            
            Point pos = path[idx];
            int dmg_taken = 0;
            
            for(const auto& tower : towers) {
                if(dist(pos, tower) <= radius) {
                    dmg_taken += damage;
                }
            }
            
            hp -= dmg_taken;
            if(hp <= 0) break;
        }
        
        if(hp > 0) return false;
    }
    return true;
}

// Helper to generate combinations
void combinations(int n, int k, int offset, vector<int>& current, vector<vector<int>>& result) {
    if (k == 0) {
        result.push_back(current);
        return;
    }
    for (int i = offset; i <= n - k; ++i) {
        current.push_back(i);
        combinations(n, k, i + 1, current, result);
        current.pop_back();
    }
}

int solve() {
    // Hardcoded setup matching Python
    vector<vector<Point>> paths;
    vector<Point> p1, p2;
    for(int i=0; i<=10; ++i) p1.push_back({0, i});
    for(int i=0; i<=10; ++i) p2.push_back({5, i});
    paths.push_back(p1);
    paths.push_back(p2);
    
    vector<Enemy> enemies = {
        {0, 20, 1},
        {1, 20, 1}
    };
    
    vector<Point> candidates = {
        {2, 5}, {0, 5}, {5, 5}
    };
    
    int radius = 3;
    int damage = 5;
    int interval = 2;
    
    int n = candidates.size();
    
    for(int k=1; k<=3; ++k) {
        vector<vector<int>> combos;
        vector<int> current;
        combinations(n, k, 0, current, combos);
        
        for(const auto& indices : combos) {
            vector<Point> selected_towers;
            for(int idx : indices) selected_towers.push_back(candidates[idx]);
            
            if(check(selected_towers, paths, enemies, radius, damage, interval)) {
                return k;
            }
        }
    }
    
    return -1;
}

int main() {
    cout << "Minimum towers needed: " << solve() << endl;
    return 0;
}

File: Day45/Part7/tower_defense.py
import itertools
import math

def solve():
    # Hardcoded Input
    # 2 Paths
    # Path 1: (0,0) -> (0,5) -> (0,10)
    # Path 2: (5,0) -> (5,5) -> (5,10)
    # Enemy 1: HP 20, Speed 1 unit/s. Path 1.
    # Enemy 2: HP 20, Speed 1 unit/s. Path 2.
    
    # Candidates:
    # 1. (2, 5) - Between paths
    # 2. (0, 5) - On Path 1
    # 3. (5, 5) - On Path 2
    
    # Tower Stats: Radius 3, Damage 5.
    # Fire Rate: Every 2s (t=0, 2, 4...)
    
    paths = [
        [(0,0), (0,1), (0,2), (0,3), (0,4), (0,5), (0,6), (0,7), (0,8), (0,9), (0,10)],
        [(5,0), (5,1), (5,2), (5,3), (5,4), (5,5), (5,6), (5,7), (5,8), (5,9), (5,10)]
    ]
    
    enemies = [
        {'path_idx': 0, 'hp': 20, 'speed': 1},
        {'path_idx': 1, 'hp': 20, 'speed': 1}
    ]
    
    candidates = [
        (2, 5),
        (0, 5),
        (5, 5)
    ]
    
    tower_radius = 3
    tower_damage = 5
    fire_interval = 2
    
    # Try k = 1, 2, 3
    for k in range(1, 4):
        for combo in itertools.combinations(candidates, k):
            # Simulate
            all_dead = True
            
            for enemy in enemies:
                path = paths[enemy['path_idx']]
                hp = enemy['hp']
                speed = enemy['speed']
                
                # Simulation
                # Time t=0, 2, 4...
                # Max time = len(path) / speed
                max_time = len(path) / speed
                
                current_hp = hp
                
                # Check damage at t=0, 2, 4...
                t = 0
                while t < max_time:
                    # Position at time t
                    # Assuming path is discrete points and speed is 1 point/s
                    # index = int(t * speed)
                    # If speed is 1, index = t
                    
                    idx = int(t * speed)
                    if idx >= len(path):
                        break # Reached end
                        
                    pos = path[idx]
                    
                    # Check towers
                    damage_taken = 0
                    for tower_pos in combo:
                        dist = math.sqrt((pos[0]-tower_pos[0])**2 + (pos[1]-tower_pos[1])**2)
                        if dist <= tower_radius:
                            damage_taken += tower_damage
                            
                    current_hp -= damage_taken
                    if current_hp <= 0:
                        break
                        
                    t += fire_interval
                
                if current_hp > 0:
                    all_dead = False
                    break
            
            if all_dead:
                return k
                
    return -1

if __name__ == "__main__":
    result = solve()
    print(f"Minimum towers needed: {result}")
